### ISSUE TITLE
Simplify creation of FixAllProvider for a Document

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn#201]
Currently, if a user needs to implement a `FixAllProvider` (or wants to for efficiency), they need to implement full support for fixing all instances of a diagnostic in three scopes: document, project, and solution. For many code fixes, changes are localized to a single document. It would be helpful to have an abstract `FixAllProvider` which allows the user to implement the "fix all" operation for a single document, but have the project and solution scopes handled by the base class (delegating handling for individual documents to the document handler as necessary).

This request is based on the challenges and concerns faced while implementing a custom `FixAllProvider` in DotNetAnalyzers/StyleCopAnalyzers#441 and DotNetAnalyzers/StyleCopAnalyzers#442, along with the feedback received in #199.

**Comment by jaredpar** [comment]
An issue to keep in mind here is performance, in particular if the implementation needs semantic information in order to function.  

Creation of a `SemanticModel` object is a *relatively* expensive operation in Roslyn.  The actual `SemanticModel` is cheap but it depends directly on `Compilation` which is *relatively* expensive to create. 

This is key when you consider having `Document` level granularity.  Every time a `Document` is modified, and hence produces a new `Document` and eventually `Solution` the current `Compilation` is invalidated.  Hence the next call to `GetSemanticModel` creates a `SemanticModel` and becomes *relatively* expensive.  

Taken to the extreme this means you will be creating a new `Compilation` for every document in the given project where this fix is applied.  In large projects this can add up very quickly.  For instance on my machine it takes ~3-5 seconds to create a `Compilation` for our Roslyn solution.  Multiply that across all of the documents in our solution and the fix quickly turns into an item that takes days to complete :smiles:

**Comment by srivatsn** [comment]
Right, when we were discussing this API, there were just too many cases where if we tried to do something smart in the API, we would end up with endless perf\correctness issues. So we chose to expose the two ends of the spectrum - the basic FixAllProvider API that makes no assumptions and the very simple BatchFixer implementation that assumes that all fixes are independent and merge cleanly. There are many variations in between and can all be built on top of the FixAllProvider API. 

The variant that @sharwell is suggesting here is interesting - this would basically be a batch fixer that assumes that fixes will merge cleanly between documents (and hence can be done without recomputing compilations or diagnostics) but not within a document and would ask the author to merge the fixes in a document. This will probably cover a lot of cases. I assume it would be `BatchDocumentFixer` or some such thing that derives from `FixAllProvider` and is exposed as a helper class. Sounds like a good idea.

**Comment by jmarolf** [comment]
@sharwell

**Comment by jaredpar** [comment]
@sharwell as @jmarolf pointed out the `BatchFixAllProvider` avoids this because it always passes `Document` objects from the original `Solution`.  This means they share the same `Compliation` object.

This overall approach works for any fix so long as:

1. Every fix only affects a single `Document` (can't sprawl across them).
2. Fixes as independent of each other.

This is true for some fixes but not other.  A base class which implemented a solution wide fix in terms of documents would have to assume the worst and this assumption would end up forcing a non-optimal creation of `Compilation` objects. 

This is actually a rather fascinating topic for me.  I recently ran into this head on in a code formatting tool I work on which was doing fixes one `Document` at a time.  The perf was fine on small projects but on large projects it was unbearable (time measured in hours).  A simple change in approach got the running time back down to the sub-minute range.  

That tool will be going OSS in the next day or two.  I'll make sure to drop a pointer here to the perf fix I made when that happens.

**Comment by sharwell** [comment]
Is there a reason the "batch documents fixer" couldn't do the same thing as the current `BatchFixAllProvider`?

**Comment by jaredpar** [comment]
@sharwell yes, what if the fixer is question wants to make changes that cross a `Document` boundary?

**Comment by sharwell** [comment]
@jaredpar I'm specifically looking for an easier way to handle cases where that's *not* the case.

**Comment by jaredpar** [comment]
@sharwell if that's not the case then have a batch solution which structures the code in the following manner:

Have a fix API in the following form:

``` csharp
abstract SyntaxNode RunFix(Document document)
```

This allows the caller to ensure the implementation won't do cross document changes.  Additionally it gives the caller implicit control over the `SemanticModel` used by the implementation.  

Now the `Solution` wide fix can be done as follows:

``` csharp

Solution RunFix(Solution originalSolution)
{
  var newSolution = originalSolution;
  foreach (var document in originalSolution.Projects.SelectMany(p => p.Documents)) 
  {
    var newRoot = RunFix(document);
    newSolution = document.WithSyntaxRoot(newRoot).Project.Solution;
  }
  return newSolution;
}
```

This ensures the `Compilation` is created at most once.

**Comment by mavasani** [comment]
@sharwell I think the best approach would be to expose APIs/components for different pieces of FixAll and allow the custom implemented FixAllProvider to either derive from some abstract type with overridable plugin points OR invoke helper APIs for default implementation of pieces that it doesn't want to customize.

FixAll has at least 3 distinct computation parts:

1. Compute applicable diagnostics to fix based on the given scope, trigger diagnostic, and trigger document/project.

2. Compute sub-fixes for the diagnostics. This could be at individual diagnostic level OR document OR project OR solution level, as appropriate for any fixer.

3. Optionally, merge sub-fixes into the final code action to apply on the solution. If step 2 generates a single fix at solution level, this is not required.

I think we should either expose an abstract FixAllProvider that implements default implementations for these components (or probably just make BatchFixAllProvider public with correct API polishing) OR expose helper APIs to invoke into these components from public FixAllProvider.

This is something that we should seriously consider fixing post 1.0, as requesting public consumers to fully implement all these 3 pieces when they just require a small tweak to the Batch fixer is probably too big an ask.

**Comment by sharwell** [comment]
As evidence of the importance of this piece of functionality, please look at DotNetAnalyzers/StyleCopAnalyzers#1382. Our reusable `DocumentBasedFixAllProvider` (*this* feature request) expedited the inclusion of a fix all provider for a code fix, offering us a 1000x performance improvement for applying this fix all operation to Roslyn.sln.

Our observations have been so consistent on this that we [modified the status page](http://stylecop.pdelvo.com/) to show an :hourglass: instead of :heavy_check_mark: for code fixes which do not use a custom document-based fix all provider.

**Comment by CyrusNajmabadi** [comment]
This has been done.  There is now [`FixAllProvider.Create`](https://github.com/dotnet/roslyn/blob/bc641a6e58a5c15ac6b8052243c2b0dd97e65e1e/src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllProvider.cs#L77) which allows fixing of documents independently and in parallel.
